windows/lcow_sandbox.go

```diff
@@ -0,0 +1,521 @@
+	scratchAttached bool           // Has a scratch been attached?
+	config          *client.Config // Represents the service VM item.
+}
+
+// startServiceVMIfNotRunning starts a service utility VM if it is not currently running.
+// It can optionally be started with a mapped virtual disk. Returns a opengcs config structure
+// representing the VM.
+func (r *LCOWRuntime) startServiceVMIfNotRunning(ctx context.Context, id string, mvdToAdd *hcsshim.MappedVirtualDisk, context string) (*serviceVMItem, error) {
+	// Use the global ID if in global mode
+	if r.globalMode {
+		id = svmGlobalID
+	}
+
+	title := fmt.Sprintf("lcowdriver: startservicevmifnotrunning %s:", id)
+
+	// Make sure thread-safe when interrogating the map
+	log.G(ctx).Debugf("%s taking serviceVmsMutex", title)
+	r.serviceVmsMutex.Lock()
+
+	// Nothing to do if it's already running except add the mapped drive if supplied.
+	if svm, ok := r.serviceVms[id]; ok {
+		log.G(ctx).Debugf("%s exists, releasing serviceVmsMutex", title)
+		r.serviceVmsMutex.Unlock()
+
+		if mvdToAdd != nil {
+			log.G(ctx).Debugf("hot-adding %s to %s", mvdToAdd.HostPath, mvdToAdd.ContainerPath)
+
+			// Ensure the item is locked while doing this
+			log.G(ctx).Debugf("%s locking serviceVmItem %s", title, svm.config.Name)
+			svm.Lock()
+
+			if err := svm.config.HotAddVhd(mvdToAdd.HostPath, mvdToAdd.ContainerPath, false, true); err != nil {
+				log.G(ctx).Debugf("%s releasing serviceVmItem %s on hot-add failure %s", title, svm.config.Name, err)
+				svm.Unlock()
+				return nil, fmt.Errorf("%s hot add %s to %s failed: %s", title, mvdToAdd.HostPath, mvdToAdd.ContainerPath, err)
+			}
+
+			log.G(ctx).Debugf("%s releasing serviceVmItem %s", title, svm.config.Name)
+			svm.Unlock()
+		}
+		return svm, nil
+	}
+
+	// Release the lock early
+	log.G(ctx).Debugf("%s releasing serviceVmsMutex", title)
+	r.serviceVmsMutex.Unlock()
+
+	// So we are starting one. First need an enpty structure.
+	svm := &serviceVMItem{
+		config: &client.Config{},
+	}
+
+	//TODO(darrenstahlmsft) add LCOW options
+	// Generate a default configuration
+	if err := svm.config.GenerateDefault(nil); err != nil {
+		return nil, fmt.Errorf("%s failed to generate default gogcs configuration for global svm (%s): %s", title, context, err)
+	}
+
+	// For the name, we deliberately suffix if safe-mode to ensure that it doesn't
+	// clash with another utility VM which may be running for the container itself.
+	// This also makes it easier to correlate through Get-ComputeProcess.
+	if id == svmGlobalID {
+		svm.config.Name = svmGlobalID
+	} else {
+		svm.config.Name = fmt.Sprintf("%s_svm", id)
+	}
+
+	// Ensure we take the cached scratch mutex around the check to ensure the file is complete
+	// and not in the process of being created by another thread.
+	scratchTargetFile := filepath.Join(r.root, scratchDirectory, fmt.Sprintf("%s.vhdx", id))
+
+	log.G(ctx).Debugf("%s locking cachedScratchMutex", title)
+	r.cachedScratchMutex.Lock()
```

Comment:
I agree that moving it to initialization makes sense, but I'd prefer to hold off on that for now to keep this similar to moby/moby LCOW code for now. Can I track this in an issue to be done later?

Reply:
Hum, maybe the moby code should be fixed ;-)

Those extra steps make the code more complicated that its need to be, how long would this need to stay the same as moby?

Also, I'm almost done with porting moby to containerd 1.0 linux wise, my next time is doing it for windows. So the LCOW code may actually disappear sooner than later from moby :innocent:

Reply:
I just don't want to diverge a ton from moby right now, as there's still lots of development going on in the equivalent moby code which needs to end up here as well. I'll take a run at changing the moby code to create the cache at init

Reply:
Honestly, I think at that point, it would be better to concentrate at least the runtime engineering effort to containerd (moby won't be make use of the snapshotter for a while longer), that would avoid duplicated effort